### PR TITLE
SAA-701 exposing the suspension details on allocations as currently not visible when pull back allocations.

### PR DESCRIPTION
**Pull request by matt-crowson-moj**
This is part one of two PR's.  Will most likely need to regenerate the types for the UI.

Ready to be reviewed but please do not merge.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/ActivitySchedule.kt [hunk 1]
```diff
@@ -190,7 +190,7 @@ data class ActivitySchedule(
   fun allocatePrisoner(
     prisonerNumber: PrisonerNumber,
     payBand: PrisonPayBand,
-    bookingId: Long?,
+    bookingId: Long,
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/ActivitySchedule.kt:193]: Booking id should never be null on allocations.  We only allocate active prisoners.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/Allocation.kt [hunk 2]
```diff
@@ -98,6 +98,13 @@ data class Allocation(
       scheduleId = activitySchedule.activityScheduleId,
       scheduleDescription = activitySchedule.description,
       isUnemployment = activitySchedule.activity.isUnemployment(),
+      deallocatedBy = deallocatedBy,
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/Allocation.kt:101]: Turns out even though deallcation details were on the model they were never exposed.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ActivityScheduleService.kt [hunk 3]
```diff
@@ -104,7 +104,7 @@ class ActivityScheduleService(
 
     schedule.allocatePrisoner(
       prisonerNumber = prisonerNumber,
-      bookingId = prisonerDetails.bookingId,
+      bookingId = prisonerDetails.bookingId ?: throw IllegalStateException("Active prisoner $prisonerNumber does not have a booking id."),
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ActivityScheduleService.kt:107]: Should never be null.  Fail fast if it is.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AppointmentService.kt [hunk 4]
```diff
@@ -45,7 +45,7 @@ class AppointmentService(
             appointmentType = AppointmentType.INDIVIDUAL,
             prisonCode = request.prisonCode,
             prisonerNumbers = listOf(it.prisonerNumber),
-            prisonerBookings = prisonerBookings.filter { entry -> entry.key == it.prisonerNumber },
+            prisonerBookings = prisonerBookings.filterKeys { k -> k == it.prisonerNumber },
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AppointmentService.kt:48]: minor refactor

File: src/main/resources/migrations/common/V2023.05.11__booking_id_mandatory_on_allocation.sql [hunk 5]
```diff
@@ -0,0 +1 @@
+ALTER TABLE allocation ALTER COLUMN booking_id SET NOT NULL;
```
Comment (matt-crowson-moj) [inline src/main/resources/migrations/common/V2023.05.11__booking_id_mandatory_on_allocation.sql:1]: Checking allocations on DEV the booking id is never null (anyway) so this migration should work fine on the existing dev db.

**Review by timharrison-moj — APPROVED** [review]
LGTM

**Pull request merged**